llvm/lib/Target/AIE/AIELiveRegs.cpp

```diff
@@ -0,0 +1,98 @@
+//===- AIELiveRegs.cpp - Liveness Analysis infrastructure ===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Implementations of the classes used to support Liveness Analysis of all
+// physical registers, including reserved registers.
+//===----------------------------------------------------------------------===//
+
+#include "AIELiveRegs.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/CodeGen/TargetSubtargetInfo.h"
+
+using namespace llvm;
+
+namespace llvm::AIE {
+
+const MachineBasicBlock *LiveRegs::popFromWorkList() {
+  const MachineBasicBlock *MBB = WorkList.front();
+  WorkList.pop();
+  WorkListSet.erase(MBB);
+
+  return MBB;
+}
+
+void LiveRegs::computeBlockLiveIns(const MachineBasicBlock *MBB,
+                                   LivePhysRegs &CurrentLive) {
+  CurrentLive.init(*TRI);
+
+  // Calculates CurrentLive by iterating over each successor of the current
+  // MBB and adding each successor's LiveIns to CurrentLive.
+  // This ensures CurrentLive is the union of LiveIns of all successor MBBs.
+  for (const MachineBasicBlock *Succ : MBB->successors()) {
+    for (MCPhysReg Reg : LiveIns[Succ]) {
+      CurrentLive.addReg(Reg);
+    }
+  }
+  // Calculates CurrentLive by bottom-up traversal of the MBB
+  for (const MachineInstr &MI : llvm::reverse(*MBB)) {
+    CurrentLive.stepBackward(MI);
+  }
+}
+
+bool LiveRegs::liveRegsAreEqual(LivePhysRegs &CurrentLive,
+                                const MachineBasicBlock *MBB) {
+  int LiveInCount = 0;
+
+  for (MCPhysReg Reg : LiveIns[MBB]) {
+    if (!CurrentLive.contains(Reg)) {
+      return false;
+    }
+    LiveInCount++;
+  }
+
+  // Updates LiveIns of MBB with CurrentLive
+  LiveIns[MBB].init(*TRI);
+  for (MCPhysReg Reg : CurrentLive) {
+    LiveInCount--;
+    LiveIns[MBB].addReg(Reg);
```

Comment:
The name of this function doesn't suggest it's updating. Also, the early return above skips updating. As far as I know this is supposed to update the liveset and indicate a change, not indicate a change and possibly update.